// @progress/kendo-react-grid/GridClientContextReader.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as t from "react";
import { useId as m, useUnstyled as R } from "@progress/kendo-react-common";
import { useLocalization as p, useInternationalization as z } from "@progress/kendo-react-intl";
import { useGridState as C } from "./GridState.mjs";
import { cloneReactElement as I } from "./utils/index.mjs";
const L = (e) => {
  const o = m(), n = e.id || o, s = R(), c = p(), i = z(), [, r] = t.useReducer((f) => !f, !0), a = t.useRef(0), l = t.useRef(0), [d, , u] = C();
  return I(e.children, {
    ...d,
    ...u,
    id: n,
    unstyled: s,
    localization: c,
    intl: i,
    scrollLeftRef: a,
    widthRef: l,
    forceUpdate: r,
    isClient: !0
  });
};
export {
  L as GridClientContextReader
};
