// @progress/kendo-react-orgchart/OrgChart.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as t from "react";
import e from "prop-types";
import { packageMetadata as s } from "./package-metadata.mjs";
import { ServerOrgChart as c } from "./server/ServerOrgChart.mjs";
import { useId as g, validatePackage as m, WatermarkOverlay as u } from "@progress/kendo-react-common";
import { ServerGroupedOrgChart as p } from "./server/ServerGroupedOrgChart.mjs";
import { ClientOrgChart as h } from "./client/ClientOrgChart.mjs";
const l = (r) => {
  const i = {
    tabIndex: 0,
    navigatable: !0,
    expandField: "expanded",
    ariaLabel: "Org Chart",
    idField: "id",
    childrenField: "items",
    avatarField: "avatar",
    avatarType: "image",
    titleField: "title",
    subtitleField: "subtitle",
    ...r
  }, {
    cardsColors: b,
    onExpandChange: F,
    onKeyDown: C,
    onItemAction: f,
    onGroupAction: y,
    onGroupBlur: v,
    onGroupFocus: O,
    id: a,
    ...n
  } = i, o = t.useRef(g()), d = !m(s, { component: "OrgChart" });
  return /* @__PURE__ */ t.createElement(h, { style: { position: "relative", ...r.style }, ...i }, r.groupField ? /* @__PURE__ */ t.createElement(p, { id: a || o.current, ...n }, r.children) : /* @__PURE__ */ t.createElement(c, { id: a || o.current, ...n }, r.children), d && /* @__PURE__ */ t.createElement(u, null));
};
l.propTypes = {
  className: e.string,
  style: e.object,
  id: e.string,
  ariaLabel: e.string,
  idField: e.string,
  childrenField: e.string,
  expandField: e.string,
  titleField: e.string,
  hasChildrenField: e.string,
  subtitleField: e.string,
  avatarField: e.string,
  cardsColors: e.array,
  data: e.array,
  groupField: e.string,
  cardHeight: e.number,
  cardWidth: e.number,
  groupTitleHeight: e.number,
  groupSubtitleHeight: e.number,
  verticalLine: e.number,
  height: e.oneOfType([e.string, e.number]),
  onExpandChange: e.func,
  onItemAction: e.func,
  onGroupAction: e.func
};
l.displayName = "KendoOrgChart";
export {
  l as OrgChart
};
